proxy/proxy.go

```diff
@@ -27,9 +28,11 @@ type roundRobinSwitcher struct {
 	index     uint32
 }
 
-func (r *roundRobinSwitcher) GetProxy(_ *http.Request) (*url.URL, error) {
+func (r *roundRobinSwitcher) GetProxy(pr *http.Request) (*url.URL, error) {
 	u := r.proxyURLs[r.index%uint32(len(r.proxyURLs))]
 	atomic.AddUint32(&r.index, 1)
+	ctx := context.WithValue(pr.Context(), colly.ProxyURLKey, u.String())
+	*pr = *pr.WithContext(ctx)
```

Comment:
Because `WithContext()` already makes a copy of `pr`, I think its better to do this without any dereferencing.

Reply:
Without doing this, I can't think of another way to update the `Context` of `pr`. Just like you said, `WithContext` only makes a copy and does not update the `Context` of `pr`.